### ISSUE TITLE
Garage Door Config

### ISSUE THREAD

**Issue by vijaydembla1** [arachnetech/homebridge-mqttthing#14]
Hi,

Please accept my sincere thanks for the work you have done, it has made me achieve my goal.
I'm writing this to discuss the issue I'm experiencing with Garage Door.

The Topic response for the status of the Garage is based no a reed switch, as a stand-alone sensor, it is working fine as the configuration of contact sensor is based on the topic response of "On"or "Off".

What is the way for me to make the script recognize the sensor response of Open and closed based on/off rather than "O ,C , o, c, S"etc.

Interestingly, the relay topic doesn't trigger the relay button, not sure what I'm missing, as the same relay is working with the Toggle function

Thanks for your help.

**Comment by i3laze** [comment]
**vijaydembla1,** please post your plugin config and tell the MQTT-device you're using.

+1 for garage setup seems really messy.

For example, GetState and SetState both use one array, but I've got two separate MQTT topics with separate statuses.
Moreover, when I set "doorValues" variable in config, MQTT logging says only a first char was published, like 'O', instead of 'ON'.

**Comment by arachnetech** [comment]
There was an error in the documentation - default target state values are O for OPEN and C for CLOSED (not 0 and 1). I'll correct this in the next release.

If you customise the values with **doorValues**, the OPEN and CLOSED target values must currently match the OPEN and CLOSED state values.

I've tested door values with strings longer than the single-character defaults and they seem to work for me. The syntax is:

```
"doorValues": [ "Open", "Closed", "Opening", "Closing", "Stopped" ]
```

Logging shows:

```
[2018-8-12 18:43:11] [Test Garage Door] Received MQTT: test/garage/current = Closed
[2018-8-12 18:43:11] [Test Garage Door] State is now: 1
[2018-8-12 18:43:29] [Test Garage Door] Received MQTT: test/garage/current = Open
[2018-8-12 18:43:29] [Test Garage Door] State is now: 0
```

MQTT logging of "State is now: 1' is referring to the internal homekit values, not the published MQTT values.

@vijaydembla1 I'm not quite sure what you mean by "the relay topic doesn't trigger the relay button, not sure what I'm missing, as the same relay is working with the Toggle function".

Note that I'm just exposing Homekit behaviour here. The supported behaviour appears to be (only) that Homekit changes the target state to open/closed to indicate whether the door should be open or closed, then the garage door can optionally send Opening/Closing followed by Stopped or Open/Closed to confirm that it has obeyed Homekit's instructions. As far as I can tell, there's no way for a garage door to tell homekit that it has been opened/closed independently of homekit.

**Comment by mientki** [comment]
Can you show your configuration for Garage Door Opener?

**Comment by arachnetech** [comment]
I was testing with:

```
    {
      "accessory": "mqttthing",
      "type": "garageDoorOpener",
      "name": "Test Garage Door",
      "url": "http://192.168.10.35:1883",
      "logMqtt": true,
      "topics":
      {
          "getCurrentDoorState":      "test/garage/current",
          "setTargetDoorState":       "test/garage/target",
          "getObstructionDetected":   "test/garage/obstruction",
          "getLockCurrentState":      "test/garagelock/current",
          "setLockTargetState":       "test/garagelock/target"
      },
      "doorValues": [ "Open", "Closed", "Opening", "Closing", "Stopped" ]
    },
```

**Comment by vijaydembla1** [comment]
{
    "accessory": "mqttthing",
    "type": "garageDoorOpener",
    "name": "TEST_Garage",
    "url": "mqtt://192.168.1.15",
    "username": "**",
    "password": "**",
    "caption": "<label (optional)>",
    "topics":
    {
        "getCurrentDoorState":      "cmnd/Sensor/POWER2",
        "setTargetDoorState":       "cmnd/Garage/POWER1",
        "getObstructionDetected":   "<topic used to get obstruction detected st$
        "getLockCurrentState":      "<topic used to get lock current state (opt$
        "setLockTargetState":       "<topic used to set lock current state (opt$
    },
    "doorValues": "O,o,,,,",
    "lockValues": "<array of 4 lock values corresponding to unsecured, secured,$
    "integerValue": "true to use 1|0 instead of true|false for obstruction dete$
},


The above is my config for the garage.
Please note: - I have also added 2 accessories, one for the door status and one to control the relay button, their topics are identical to the one above and they seem to work properly, It's just the garage is always open in homekit, when i press it, it says closing, but i don't see any effect on relay (the toggle).

Please let me know if i'm missing something.

**Comment by arachnetech** [comment]
I would recommend removing the settings with `< ... >` (`getLockCurrentState`, `setLockTargetState`, `lockValues`, `integerValue`) as they might confuse things. Those are just intended as comments in the ReadMe.

As far as I can tell, HomeKit's Garage Door (like it's Security System) mandates that *HomeKit* controls the state that it should be set to, and the door is expected to respond once it's obeyed the command to confirm that HomeKit's state has been reached. It may be possible to set up your switch by setting up and automation in homekit to make the switch change the target state of the garage door. However, you can't just tell homekit that the state has changed, because it will always assume that the door should be moving to homekit's target state.

**Comment by Kepete** [comment]
@vijaydembla1 As I see you are also trying to use a esp flashed with tasmota for your garage door. Could I ask what setup are you running? 
I have connected two contact sensors and a relay to the tasmota. Now I would like to have a closed mqtt topic and a separate open mqtt topic. Becaouse tasmota gives me the state of one contact sensor and the state of the other.

**Comment by mientki** [comment]
I also want to do just like you. Two reed sensors + relay on tasmota. Can you help?

**Comment by arachnetech** [comment]
Could you publish to the same MQTT topic from both contact sensors when they transition as follows:

When closed sensor no contact -> opening
When open sensor contact -> open
When open sensor no contact -> closing
When closed sensor contact -> closed

**Comment by TLCary** [comment]
For some reason HomeKit transitions to Opening from Open regardless of the MQTT State it receives.  Open should transition to Closing, of course.  I've tried it by feeding the default values directly to MQTT, with custom statuses, etc...  it always goes to Opening.  Odd.   Like the custom feature.  TASMOTA parses everything to all caps so you have to differentiate outside of case.  Love the idea of one plugin to address all these services... please keep up the good work.

**Comment by brefra** [comment]
> I would recommend removing the settings with `< ... >` (`getLockCurrentState`, `setLockTargetState`, `lockValues`, `integerValue`) as they might confuse things. Those are just intended as comments in the ReadMe.
> 
> As far as I can tell, HomeKit's Garage Door (like it's Security System) mandates that _HomeKit_ controls the state that it should be set to, and the door is expected to respond once it's obeyed the command to confirm that HomeKit's state has been reached. It may be possible to set up your switch by setting up and automation in homekit to make the switch change the target state of the garage door. However, you can't just tell homekit that the state has changed, because it will always assume that the door should be moving to homekit's target state.

Pull request #28 contains the fix to have the GarageDoor device in HomeKit in sync when the door has been controlled from outside of Homebridge.

**Comment by arachnetech** [comment]
My <comments in settings> have caused a certain amount of confusion, so I've started changing the format. As discussed in the pull request, your suggestions were very helpful. I was wrong in my belief that target state couldn't be changed outside HomeKit. The Home App doesn't like to show it when it changes, but closing and re-opening it does show a new externally-set target state. Version 1.0.15 now supports external setting of target states for garage door and security system.

**Comment by transistorgit** [comment]
Hi, I'm using MQTTThing for all my devices, great!
Now I'm switching over my MQTT-controlled gateway door (same as a garage door) to use also mqttthing.
It works nice so far, but it seems that the "opening"/"closing" states are interchanged in the homekit view, when the door is triggered externally. When triggered by homekit, all states are showing correct order: Door open - touch it - closing - closed. But when its open an I hit the external button, homekit shows "opening" and than "closed" as the final state. The same when it is closed. Hitting the external button - homekit shows "closing", final state is open.

I checked the Mqtt traffic in parallel. It looks completely correct. TargetDoorState shows "Closing", but homekit shows "opening" at the same time (when triggered externally, as said).

Here's my config:
      {
        "accessory": "mqttthing",
        "type": "garageDoorOpener",
        "name": "Toreinfahrt",
        "url": "http://192.168.168.112:1883",
        "topics":
        {
          "setTargetDoorState":       "Toreinfahrt/Target",
          "getTargetDoorState":       "Toreinfahrt/TargetFeedback",
          "getCurrentDoorState":      "Toreinfahrt/Status",
          "setLockTargetState":       "Toreinfahrt/LockTarget",
          "getLockTargetState":       "Toreinfahrt/unused",
          "getLockCurrentState":      "Toreinfahrt/unused",
          "getObstructionDetected":   "Toreinfahrt/unused"
        },
        "doorValues": [ "Open", "Closed", "Opening", "Closing", "Stopped" ],
        "lockValues": [ "Unsecured", "Secured", "Jammed",  "Unknown" ]
       }

TargetFeedback will only show the values "Open" and "Closed".
Status will show all states "Open", Closed", "Opening", "Closing", "Stopped" exactly written like that.

By the way: homekit doesn't show "stopped", but switches to "Open". Is this Homekit behaviour?

Please tell me what I can do to resolve this, or if I should open a new issue.
best regards

**Comment by arachnetech** [comment]
The Home App doesn’t appear to handle changes in target state correctly. It seems to read the target state when it is opened but then not update it subsequently. If the state then changes to something other than this _initial_ target state, the Home App then shows that the accessory is ‘moving towards’ the initial target state that it read (when really it is moving in the other direction). When I first saw this behaviour, I assumed that target state could only be changed by HomeKit itself. However, if you close then re-open the Home App after changing the target state, it shows the correct state.

**Comment by transistorgit** [comment]
Ok, that explains why I also saw correct target states sometimes. Hopefully Apple will fix that at some time. Thank you and have a nice sunday!

**Comment by TLCary** [comment]
"I don't think so Tim."   I use MQTTThing for everything but the Garage Door. I would like to use it for everything, but..  I use mqttgaragedoor because it works just fine.  It's not Apple or the Home App that isn't responding to changes in the device from external inputs.

**Comment by arachnetech** [comment]
I need to try this mqttgarage door...

**Comment by arachnetech** [comment]
Reopening to remind me to look at https://github.com/iomax/homebridge-mqttgaragedoor.

**Comment by peteroz12** [comment]
Hi there, I'm trying to get my Garadget (https://community.garadget.com/t/mqtt-support/3226/5) to work, it seems to open but not close or get the correct status. My config is:

{
    "bridge": {
        "name": "Ozone Garage",
        "username": "XX:XX:XX:XX:XX:XX",
        "port": XXXXX,
        "pin": "XXX-XX-XXX"
    },
    "description": "Testing Garage Door",
    "platforms": [],
    "accessories": [{
        "accessory": "mqttthing",
        "type": "garageDoorOpener",
        "name": "Garage Door",
        "url": "http://XXX.XXX.XXX.XXX:1883",
        "logMqtt": true,
        "topics": {
            "setTargetDoorState": "garadget/Door/command",
            "getTargetDoorState": "garadget/Door/status",
            "getCurrentDoorState": "garadget/Door/command",
            "getObstructionDetected": "garadget/Door/status"
        },
        "doorValues": ["open", "closed", "opening", "closing", "stopped"]
    }]
}

Is there something I'm doing wrong?

**Comment by TLCary** [comment]
Invalid JSON at Port.

Look for invalid characters.  A common problem is using an editor that helps you out by automatically making a “closing parentheses”.  That’s a different character than the " that JSON expects.  

Jasonlint.com  or https://codebeautify.org/jsonvalidator etc

**Comment by pasna** [comment]
Hi, Best wishes for the coming New Year 2019 everyone. Your plugin is great and I just complete configure all my lightings, I have tested, they work great, I really appreciate your hard work. Now I am working on using Sonoff SV flashed Tasmota 6.4.1, I configure the device as follow -
GPIO5 -> Switch2 connect to reed switch - one reed switch only.

PulseTime1 10
SwitchMode1 1
SwitchMode2 2
PowerOnState 0
PowerRetain 1
SwitchTopic 0

status read -
20:08:24 MQT: stat/frontgatetest/STATUS = {"Status":{"Module":3,"FriendlyName":["Front_Gate_Test"],"Topic":"frontgatetest","ButtonTopic":"0","Power":0,"PowerOnState":0,"LedState":1,"SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[1,2,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":1}}

rule1 read -
on switch2#state=1 do publish test/garage/current %value% endon on switch2#state=0 do publish test/garage/current o endon

config file -
        {
            "accessory": "mqttthing",
            "type": "garageDoorOpener",
            "name": "Front Gate Test",
            "url": "http://192.168.0.55:1883", 
            "username": "my_name",
            "password": "my_password",
            "logMQtt": true,
            "caption": "Front Gate Test",
            "topics": {
                "setTargetDoorState": "cmnd/frontgatetest/POWER",
                "getTargetDoorState": "test/garage/current",
                "getCurrentDoorState": "test/garage/current"
            },
            "doorValues": [
                "1",
                "1",
                "o",
                "c",
                "S"
            ],
            "lockValues": []
        }
with is setting, I am able to get the device trigger 1 second pulse and control my sliding gate to open and close without any problem regardless of how Home App display status. However I felt that this plugin is fascinating and I love to use fewer plugins. At current status these cases were what I had faced.
pt 1) I cannot use other doorValues than what shows above otherwise sonoff will not trigger, I found that "1" will command sonoff to trigger,  "o" , "c", "S" did not push Home App to reflect status correctly. The app will spin forever and showing either opening or closing.

pt 2) I have to configure GPIO5 -> Switch2 and put reed switch into the play, reed switch does not control any relay. With reed switch and rule1 I can put the app to come to "Closed" position ( On Home App ).  I attach this reed switch to the gate and I got notification pop-up on iPhone when gate is opened or closed and it is perfect. However when I manually, without push on the App Icon, opening the gate and the reed switch move apart, Home App showing closing instead of opening ! I wish it shows "Open" instead.

pt 3) When I restart Homebridge, the plugin fail to retain LWT and the Icon show "Open" but does not trigger relay. This only affect the garage door icon not the lightings.


Have a great holiday.

**Comment by vijaydembla1** [comment]
Hi All,

I managed to get it to work, all i need is the pulse time to be included. Below is my current config:-

{
    "accessory": "mqttthing",
    "type": "garageDoorOpener",
    "name": "garage-test",
    "url": "mqtt://192.168.1.15",
    "username": "pi",
    "password": "WWWWWWWW",
    "caption": "test-garage",
    "topics":
    {
        "setTargetDoorState":       "cmnd/Garage/power3",
        "getTargetDoorState":       "cmnd/2/POWER2",
        "getCurrentDoorState":      "cmnd/2/POWER2"
    },
    "doorValues": ["OFF", "ON" ,"o" ,"c" ,"S" ]
}


I have 4 relays and 1 reed switch connected. I need help in adding pulse time to relay 3 only (garage/power3.

I have been able to add all of them and with added pluse time, i'll be able to set up my garage. 
Sorry i have been away and it took ages for me to recall what i orginially did.

Thanks again

**Comment by NorthernMan54** [comment]
If your config is good, could you add it to the Wiki of Test Configurations so others can make use of it?

https://github.com/arachnetech/homebridge-mqttthing/wiki/Tested-Configurations

**Comment by vijaydembla1** [comment]
The config needs 2 improvements:-
a) The relay should be timed for a second or so
b) after rebooting either: RaspberryPi, Homebridge. ESP8266
The garage status shows open. You need to physically seperate the reed switch and bring them together to see the correct status in homekit.

Once the above 2 issues have been addressed, then i can upload it to the above link and with various others

Developer of this code might need to help on this :-)

Once again, thanks to all

**Comment by tobekas** [comment]
please try Version 1.0.42

**Comment by arachnetech** [comment]
Closing old issue.
